ath79: fortinet: convert to nvmem-layout

Leo-PL opened the pull request:
This series converts both FAP-220-B and FAP-221-B models to use `nvmem-layout` for parsing MAC addresses, so no userspace scripts are needed to configure them. While at that, the nvmem-layout is consolidated in the common part.

Run tested on FAP-220-B.

Leo-PL committed:
ath79: fortinet-fap-220-b: convert to nvmem-layout

Now that MAC address parser supports the hex format (without
delimiters), use the canonical MAC address stored in U-boot partition.
Get rid of the "mac-address-increment" binding.
While at that, convert ART partition too.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
ath79: fortinet-fap-221-b: convert to nvmem-layout

Now that MAC address parser supports the hex format (without
delimiters), use the canonical MAC address stored in U-boot partition.
Get rid of the userspace adjustments which are no longer necessary.
While at that, move the mac-base to the common part, as it is again
exactly the same in both models.

And convert ART partition too - keep that one separate, as calibration
data length differs between the models.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL commented:
@mpratt14 please double-check if I hadn't swapped the MACs for the radios on FAP-221-B by accident - as I'm not sure which PHY ends up the first one, and both are configured according to `PHYNBR`.

mpratt14 commented:
I wouldn't bother calling the nodes "mac-address-hex", it should be clear by the fact that it's length 0xc

mpratt14 commented:
I did

eth.     +0
phy0.  +1
phy1.  +8

mpratt14 commented:
usually PCI is the phy0 I think....

Leo-PL commented:
On other multi-ath9k devices, the AHB is usually the phy0, then the PCIe ones, So what I defined in .dts should be correct.

neheb commented on target/linux/ath79/dts/ar7161_fortinet_fap-220-b.dts:
```diff
@@ -154,9 +152,16 @@
 	#address-cells = <1>;
 	#size-cells = <1>;
```
Not needed,

neheb commented:
Might as well convert art too.

neheb commented on target/linux/ath79/dts/arxxxx_fortinet_loader.dtsi:
```diff
@@ -40,6 +40,21 @@
 				label = "u-boot";
 				reg = <0x000000 0x040000>;
 				read-only;
+				compatible = "nvmem-cells";
```
space after read-only.

neheb replied:
actually nvm. nvmem-cells is not needed anymore.

Leo-PL replied:
So I can drop lines 43-45?

neheb replied:
yeah. See https://github.com/neheb/openwrt/commit/a7bfe15a3b675259835dd5271e2886b8ff010ffd for examples.

mpratt14 commented:
why is is ok to remove "nvmem-cells" compatible?

i think we should wait until fw_devlink is fixed before we try to remove that

neheb commented:
Is it even needed? Works without locally.

mpratt14 commented:
who knows if it stops working if nvmem gets restructured upstream. I'm just of the opinion that the dts should be as formal as possible, we already have the "bad cell count" warning just because we don't put enough #address-cells and #size-cells in dts

neheb commented:
In my PRs, I’ve fixed all such data warnings. nvmem-cells should not be here.

mpratt14 commented:
ok, I see the kernel has a bool called `add_legacy_fixed_of_cells` which is triggered by the `nvmem-cells` compatible, so I agree with removing it 

(although, I wish they didn't do it this way as it used to just be the way to generally declare a node as being used by the nvmem driver)

Leo-PL commented:
Okay, I'll remove those lines, might tackle `art` too. I'll need to retest of course.

Leo-PL commented:
Allright, dropped the `-hex` suffix and nvmem-cells property - it wasn't needed anyway since there were no sub-nodes directly referencing the MAC, only one under nvmem-layout node.

mpratt14 commented:
you were right in that phy0 is the WMAC device and phy1 is PCI

on the 221b, the mac increments were 1 and 8 though

Leo-PL commented:
Just checked the diff to be extra sure - they still are.

mpratt14 commented:
how about your 220b board, aren't they the same?

Leo-PL commented:
Just checked boot logs from factory FW and it's even weirder. The offsets are... 9 and 2. In that order, taking PCI devices ordering into account. Time to fix that.

Leo-PL committed:
ath79: fortinet-fap-220-b: fix WLAN MAC addresses

Addresses were swapped compared to the factory firmware. In addition to
that, one of them was shifted by -1. Fix that by setting wlan0 MAC
offset to 9, and wlan1 MAC offset to 2.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

mpratt14 commented:
if it were not for a bunch of exceptions, the kernel would indeed probe all devices in the order they appear in dts under the bus

some drivers force probing to happen late by setting it to be called by a later initcall like late_initcall. If I had to guess that probably applies to one of them

there's also the fw_devlink feature which we have currently disabled but I'm trying to fix which also controls probe order

mpratt14 commented:
> Just checked boot logs from factory FW and it's even weirder. The offsets are... 9 and 2. In that order, taking PCI devices ordering into account.

for all I know mine is flipped too, but I never bother with that fine detail since phy0 and phy1 have been flipped in the past between kernel versions, it could happen again...

which log do you look at? I never really see a factory boot log at all, it's just blank after the bootloader until busybox starts

mpratt14 commented:
there is still a diff in the last commit that should be in the previous one

where you got rid of "mac-address-hex" for 220b, that's in the commit for 221b

Leo-PL commented:
> there is still a diff in the last commit that should be in the previous one
> 
> where you got rid of "mac-address-hex" for 220b, that's in the commit for 221b

It was only in one of previous patchsets, never in upstream. Changed that according to @neheb's suggestion, the node is simply called `mac-address` now and it's in the common part.

> which log do you look at? I never really see a factory boot log at all, it's just blank after the bootloader until busybox starts

I saved some console output after playing with ifconfig and exploring procfs and sysfs - it was there.
```
ifconfig
br0         If index:  5 Link encap:Ethernet  HWaddr 00:09:0F:49:01:3E  
            inet addr:192.168.1.2  Bcast:192.168.1.255  Mask:255.255.255.0
            UP BROADCAST NOTRAILERS RUNNING MULTICAST  MTU:1500  Metric:1
            RX packets:9 errors:0 dropped:0 overruns:0 frame:0
            TX packets:39 errors:0 dropped:0 overruns:0 carrier:0
            collisions:0 txqueuelen:400 
            RX bytes:1345 (1.3 KiB)  TX bytes:18963 (18.5 KiB)

eth1        If index:  3 Link encap:Ethernet  HWaddr 00:09:0F:49:01:3E  
            UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
            RX packets:10 errors:0 dropped:0 overruns:0 frame:0
            TX packets:39 errors:0 dropped:0 overruns:0 carrier:0
            collisions:0 txqueuelen:1000 
            RX bytes:1614 (1.5 KiB)  TX bytes:18963 (18.5 KiB)

lo          If index:  1 Link encap:Local Loopback  
            inet addr:127.0.0.1  Mask:255.0.0.0
            UP LOOPBACK RUNNING  MTU:16436  Metric:1
            RX packets:0 errors:0 dropped:0 overruns:0 frame:0
            TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
            collisions:0 txqueuelen:0 
            RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

wifi0       If index:  7 Link encap:IEEE80211  HWaddr 00:09:0F:49:01:40  
            UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
            RX packets:4084 errors:0 dropped:0 overruns:0 frame:0
            TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
            collisions:0 txqueuelen:539 
            RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
            Interrupt:49 Memory:b0010000-b0020000 

wifi1       If index:  6 Link encap:IEEE80211  HWaddr 00:09:0F:49:01:47  
            UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
            RX packets:3502 errors:0 dropped:0 overruns:0 frame:0
            TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
            collisions:0 txqueuelen:539 
            RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
            Interrupt:48 Memory:b0000000-b0010000
```

mpratt14 commented:
> the node is simply called `mac-address` now and it's in the common part

yes you did, but the changes are split between 2 commits

look at the third one carefully, search "220"

https://github.com/openwrt/openwrt/pull/13997/commits/06cb4aa7314e7a55d37b0f3eab51066b131e3695.patch

Ansuel commented:
Soo is this ready? Should we backport the first commit to stable?

mpratt14 commented:
@Leo-PL basically you need to do the following

```
git reset --soft HEAD^

# note that both dts have changes...
git status

# unstage everything
git status | grep modified | cut -d ':' -f 2 | xargs git restore --staged

# amend previous commit with 220 dts
git add target/linux/ath79/dts/ar7161_fortinet_fap-220-b.dts
git add target/linux/ath79/dts/arxxxx_fortinet_loader.dtsi
git commit --amend

# re-apply the rest into the last commit
git add target/linux/ath79/*
git commit -c ORIG_HEAD
```

Leo-PL commented:
> > the node is simply called `mac-address` now and it's in the common part
> 
> yes you did, but the changes are split between 2 commits
> 
> look at the third one carefully, search "220"
> 
> https://github.com/openwrt/openwrt/pull/13997/commits/06cb4aa7314e7a55d37b0f3eab51066b131e3695.patch

@mpratt14  Okay, I see now. Good catch. I'll fix that right away - I prefer the `git rebase --interactive` approach.

@Ansuel - the device wasn't backported to 23.05 yet, but after this one gets merged - why not.

Leo-PL commented:
Okay, now it should be ready. @Ansuel, @mpratt14, @neheb please re-review.

mpratt14 commented:
still there...haha

mpratt14 commented:
try again

we want changes to
220-b.dts
and 
loader.dtsi
in the middle commit instead of the last one

Leo-PL commented:
Okay, I had a bit different vision here - I converted 220 separately first, so changes for 221 and the common part could be reverted easily if needed, without touching 220. But I can extract the common part first.

mpratt14 commented:
I understand the concern about reverting, but this is a pretty safe change to make, and reverts don't need to be a perfect reverse of the original.

I think it's more important that checking out each commit one at a time results in successful build (which is also not *always* possible but a good goal)

even more basic than that, is to make sure the commit changes always match the commit title

anyway good job 👍
